src/main/java/codeu/model/store/basic/MessageStore.java

```diff
@@ -79,10 +81,23 @@ public void addMessage(Message message) {
   public List<Message> getMessagesInConversation(UUID conversationId) {
 
     List<Message> messagesInConversation = new ArrayList<>();
+    HashSet<UUID>idSet = new HashSet<UUID>();
 
     for (Message message : messages) {
       if (message.getConversationId().equals(conversationId)) {
-        messagesInConversation.add(message);
+        if (!idSet.contains(message.getId())) {
+          idSet.add(message.getId());
+          messagesInConversation.add(message);
+        } else {
+          for (Message old : messagesInConversation) {
+            if (old.getId().equals(message.getId())
+                && old.getCreationTime().isBefore(message.getCreationTime())) {
+              messagesInConversation.remove(old);
+              messagesInConversation.add(message);
+              break;
+            }
```

Comment:
Hm, this is a bit confusing to me, here's what I'd do instead:

```java
private List<Message> getMessagesInConversation(UUID conversationId) {
return messages.stream().filter(message -> 
    message.getConversationId().equals(conversationId)).collect(Collectors.toList());
}

public List<Message> getCurrentMessagesInConversation(UUID conversationId) {
    List<Message> messages = getMessagesInConversation(conversationId);
    // Ensures messages are sorted in reverse chronological order (newest -> oldest): Please check if they are 
    // somewhere before this. if they are, can remove
    Collections.sort(messages, (o1, o2) -> o2.getCreationTime().compareTo(o1.getCreationTime()));
    // Put into a tree set to get rid of all duplicate entries after the newest message (comparator looks at UUID 
    // to do this), and then collects back into list
    return messages.stream().collect(collectingAndThen(toCollection(() -> new TreeSet<> . 
    (Comparator.comparing(UUID::toString))), ArrayList::new));
}
```

Let me know if I misunderstood your code. double check this for me bc I didn't test it